apps/emqx_gcp_device/src/emqx_gcp_device.erl

```diff
@@ -0,0 +1,268 @@
+-type registry() :: binary().
+-type device_loc() :: {project(), location(), registry()}.
+-type key_type() :: binary().
+-type key() :: binary().
+-type expires_at() :: pos_integer().
+-type key_record() :: {key_type(), key(), expires_at()}.
+-type created_at() :: pos_integer().
+-type extra() :: map().
+
+-record(emqx_gcp_device, {
+    id :: deviceid(),
+    keys :: [key_record()],
+    device_loc :: device_loc(),
+    created_at :: created_at(),
+    extra :: extra()
+}).
+-type emqx_gcp_device() :: #emqx_gcp_device{}.
+
+-type formatted_key() ::
+    #{
+        key_type := key_type(),
+        key := key(),
+        expires_at := expires_at()
+    }.
+-type encoded_config() :: binary().
+-type formatted_device() ::
+    #{
+        deviceid := deviceid(),
+        keys := [formatted_key()],
+        config := encoded_config(),
+        project => project(),
+        location => location(),
+        registry => registry(),
+        created_at => created_at()
+    }.
+-export_type([formatted_device/0, deviceid/0, encoded_config/0]).
+
+-define(TAB, ?MODULE).
+
+-dialyzer({nowarn_function, perform_dirty/2}).
+
+%%--------------------------------------------------------------------
+%% API
+%%--------------------------------------------------------------------
+
+-spec put_device(formatted_device()) -> ok.
+put_device(FormattedDevice) ->
+    try
+        put_device_no_transaction(FormattedDevice)
+    catch
+        _Error:Reason ->
+            ?SLOG(error, #{
+                msg => "Failed to put device",
+                device => FormattedDevice,
+                reason => Reason
+            }),
+            {error, Reason}
+    end.
+
+-spec get_device(deviceid()) -> {ok, formatted_device()} | not_found.
+get_device(DeviceId) ->
+    case ets:lookup(?TAB, DeviceId) of
+        [] ->
+            not_found;
+        [Device] ->
+            {ok, format_device(Device)}
+    end.
+
+-spec remove_device(deviceid()) -> ok.
+remove_device(DeviceId) ->
+    ok = mria:dirty_delete({?TAB, DeviceId}),
+    ok = put_config(DeviceId, <<>>).
+
+-spec get_device_actual_keys(deviceid()) -> [key()] | not_found.
+get_device_actual_keys(DeviceId) ->
+    try ets:lookup(?TAB, DeviceId) of
+        [] ->
+            not_found;
+        [Device] ->
+            actual_keys(Device)
+    catch
+        error:badarg ->
+            not_found
+    end.
+
+-spec import_devices([formatted_device()]) ->
+    {NumImported :: non_neg_integer(), NumError :: non_neg_integer()}.
+import_devices(FormattedDevices) when is_list(FormattedDevices) ->
+    perform_dirty(fun() -> lists:foldl(fun import_device/2, {0, 0}, FormattedDevices) end).
+
+%%--------------------------------------------------------------------
+%% Internal API
+%%--------------------------------------------------------------------
+
+-spec create_table() -> ok.
+create_table() ->
+    ok = mria:create_table(?TAB, [
+        {rlog_shard, ?AUTH_SHARD},
+        {type, ordered_set},
```

Comment:
Can the type be  `set`?

Reply:
Ack.

Reply:
From past discussions, we try to use `ordered_set` since it has the same properties of `set` without too much of a performance penalty.

Reply:
OK. So should we keep `ordered_set` here, right?
Also, just to know: what was the reason to try to use `ordered_set`? 🫣